packages/compass-aggregations/src/modules/pipeline.ts

```diff
@@ -320,33 +320,40 @@ const moveStage = (state: State, action: AnyAction): State => {
 const selectStageOperator = (state: State, action: AnyAction): State => {
   const operatorName = action.stageOperator;
   const oldStage = state[action.index];
-  if (operatorName !== oldStage.stageOperator) {
-    const newState = copyState(state);
-    // if the value of the existing state operator has not been modified by user,
-    // we can easily replace it or else persist the one user changed
-    let value = getStageDefaultValue(operatorName, action.isCommenting, action.env);
-    if (hasUserChangedStage(oldStage, action.env)) {
-      value = oldStage.stage;
-    }
-    newState[action.index].stageOperator = operatorName;
-    newState[action.index].stage = value;
-    newState[action.index].isExpanded = true;
-    newState[action.index].isComplete = false;
-    newState[action.index].previewDocuments = [];
-    if (
-      [SEARCH, SEARCH_META, DOCUMENTS].includes(newState[action.index].stageOperator) &&
-      action.env !== ADL && action.env !== ATLAS
-    ) {
-      newState[action.index].isMissingAtlasOnlyStageSupport = true;
-    } else {
-      newState[action.index].isMissingAtlasOnlyStageSupport = false;
-    }
-    const { isValid, syntaxError } = validateStage(newState[action.index]);
-    newState[action.index].isValid = isValid;
-    newState[action.index].syntaxError = syntaxError;
-    return newState;
+
+  if (operatorName === oldStage.stageOperator) {
+    return state;
   }
-  return state;
+
+  // if the value of the existing state operator has not been modified by user,
+  // we can easily replace it or else persist the one user changed
+  let value;
+  if (hasUserChangedStage(oldStage, action.env)) {
+    value = oldStage.stage;
+  }
+  else {
+    value = getStageDefaultValue(operatorName, action.isCommenting, action.env);
+  }
+
+  const newState = copyState(state);
+
+  newState[action.index].stageOperator = operatorName;
+  newState[action.index].stage = value;
+  newState[action.index].isExpanded = true;
+  newState[action.index].isComplete = false;
+  newState[action.index].previewDocuments = [];
+  newState[action.index].isMissingAtlasOnlyStageSupport = !!(
+    [SEARCH, SEARCH_META, DOCUMENTS].includes(operatorName) &&
+    action.env !== ADL && action.env !== ATLAS
+  );
+
+  const { isValid } = validateStage(newState[action.index]);
+  newState[action.index].isValid = isValid;
+  // Clear the error message because the user just changed the operator. Even
+  // though the text wouldn't have changed if the user has edited it.
+  newState[action.index].syntaxError = null;
```

Comment:
That's literally the ticket, though 😆 

I'm also not sure about this entire thing.

Reply:
I think the problem stems from us not replacing the stage text once the user has edited it.

Reply:
we show the warning in case we are not able to parse the contents of stage (which is independent of stage operator). so changing between stage operators with invalid content, it's still invalid stage, right? (but we also hide it as preferred)

but the error message is tied to the stage operator and we should be clearing that as user changes between operators.

Reply:
Aah sorry I mixed up syntaxError and error. Didn't realise that's a different error var. Changed now.